### ISSUE TITLE
Loading of shuffle buffer occurs every epoch drastically increasing training time using tf.datasets

### ISSUE THREAD

**Issue by LastHorizon** [tensorflow/tensorflow#47090]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 4.0.2
- Python version: 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 8.0.2.39
- GPU model and memory: 2x RTX 3090 w/24GB memory

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

Hi all,

Firstly, apologies if this is in the wrong place but I feel this is something that would be amazing to address if possible!
I am training a model using images stored in a directory and creating datasets with the tf.keras.preprocessing.image_dataset_from_directory() method. Images are set as 50 x 50 pixels, greyscale.
My training routine is lightning fast thanks to the RTX 3090 graphics card. However, there is substantial I/O bottlenecking when filling the shuffle buffer up. Is there a way to address this and make it faster?

I can see the shuffle buffer being reloaded every single Epoch which takes significant time.

Thanks!

**Describe the expected behavior**

Loading of the shuffle buffer only once to significantly decrease training time.

**Standalone code to reproduce the issue**

```

## create data sets 

        train_ds = tf.keras.preprocessing.image_dataset_from_directory(
          data_dir,
          validation_split=validation_split,
          subset="training",
          seed=seed,
          image_size=(img_height, img_width),
          batch_size=batch_size,
          color_mode='grayscale'
        )

        val_ds = tf.keras.preprocessing.image_dataset_from_directory(
          data_dir,
          validation_split=validation_split,
          subset="validation",
          seed=seed,
          image_size=(img_height, img_width),
          batch_size=batch_size,
          color_mode='grayscale'
        )

## create and compile model 
model = get_compiled_model()

## train network (call back simply saves epoch data if epoch result is better than previous)
        history = model.fit(
          train_ds,
          validation_data=val_ds,
          epochs=epochs,
          initial_epoch=offset,
          callbacks=callbacks,
          steps_per_epoch=None,
          validation_steps=None
        )
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@LastHorizon 
I ran the code shared and face a different error, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/05edbb8ec298189d44c346d9568aaf1f/untitled529.ipynb).
Please share code that is indented and provide all dependencies, if possible share a colab gist as well with the data n error reported.
